### PR TITLE
Implement StateManager (fixes #150)

### LINKED ISSUE #150: LabeledEnum helper property to replace Docflow
`LabeledEnum` is now consistently used in HasGeek apps for status flag values. With the support for grouped values in #148, it now replaces the [workflow state definition feature in Docflow](http://docflow.readthedocs.io/en/latest/).

LabeledEnum version:

```python
class POSTSTATUS(LabeledEnum):
    DRAFT =        (0,  __("Draft"))         # Being written
    PENDING =      (1,  __("Pending"))       # Pending email verification
    CONFIRMED =    (2,  __("Confirmed"))     # Post is now live on site
    REVIEWED =     (3,  __("Reviewed"))      # Reviewed and cleared for push channels

    UNPUBLISHED = {DRAFT, PENDING}
    LISTED = {CONFIRMED, REVIEWED}
```

Docflow version:
```python
class PostStatus(DocumentWorkflow):
    state_attr = 'status'

    draft     = WorkflowState(0, title="Draft",     description="Only owner can see it")
    pending   = WorkflowState(1, title="Pending",   description="Pending review")
    confirmed = WorkflowState(2, title="Published", description="Published")
    reviewed  = WorkflowState(3, title="Withdrawn", description="Withdrawn by owner")

    unpublished = WorkflowStateGroup([draft, pending], title="Not Published")
    listed = WorkflowStateGroup([confirmed, reviewed], title="Listed")
```

`LabeledEnum`'s state values are lower level than the model that requires state values, whereas `DocumentWorkflow` is a wrapper around the model and thus can't supply data essential to the model itself.

However, Docflow provides two other features that `LabeledEnum` currently can't help with. Coaster should add support:

## State queries

Consider how status queries are done with `LabeledEnum` ([from Hasjob's `JobPost` model](https://github.com/hasgeek/hasjob/blob/01a4a2f9fee24873c5f7e3199911e254f46464c7/hasjob/models/jobpost.py#L236-L243)):

```python
class JobPost(BaseMixin, db.Model):
    status = db.Column(db.Integer, nullable=False, default=POSTSTATUS.DRAFT)

    def is_draft(self):
        return self.status == POSTSTATUS.DRAFT

    def is_pending(self):
        return self.status == POSTSTATUS.PENDING

    def is_unpublished(self):
        return self.status in POSTSTATUS.UNPUBLISHED
```
Contrast with Docflow:

```python
>>> wf = PostStatus(doc1)
>>> # First method, using the `is` operator
>>> wf.state is wf.draft
True
>>> # Second method, by calling the state to enquire if it's current
>>> wf.draft()
True
>>> wf.pending()
False
```

`LabeledEnum` requires a profusion of explicitly declared helper methods, while `DocumentWorkflow` constructs them automatically. However, `DocumentWorkflow` requires the document to be wrapped in the workflow first, which imposes the additional requirement of having access to the wrapper (for example in a template, which has the object it receives and everything within it, but nothing external).

It'll be nice if `LabeledEnum` came with a helper object that worked something like this:

```python
class JobPost(BaseMixin, db.Model):
    status = LabeledEnumProperty(POSTSTATUS, db.Column(db.Integer, nullable=False, default=POSTSTATUS.DRAFT))
    …

>>> post = JobPost()
>>> post.status
0
>>> post.status == POSTSTATUS.DRAFT
True
>>> post.status.DRAFT
True
>>> post.status.UNPUBLISHED
True
>>> post.status.PENDING
False
>>> JobPost.query.filter(JobPost.status.DRAFT).all()  # Expands to JobPost.status == POSTSTATUS.DRAFT
>>> JobPost.query.filter(JobPost.status.UNPUBLISHED).all()  # Expands to JobPost.status.in_(POSTSTATUS.UNPUBLISHED)
```

## State transitions

The second feature that Docflow provides is state transitions, as in this example:

```python
class PostStatus(DocumentWorkflow):
    …
    @draft.transition(pending, None, title='Submit')
    def submit(self):
        """
        Change workflow state from draft to pending.
        """
        # Do something here
        # ...
        pass  # State will be changed automatically if we don't raise an exception
```

In this case the `LabeledEnumProperty` object can provide a `transition` wrapper:

```python
class JobPost(BaseMixin, db.Model):
    status = LabeledEnumProperty(POSTSTATUS, db.Column(db.Integer, nullable=False, default=POSTSTATUS.DRAFT))

    @status.transition(POSTSTATUS.DRAFT, POSTSTATUS.PENDING)
    def submit(self):
        """
        Change status from draft to pending
        """
        self.datetime = func.utcnow()
```

### PR DESCRIPTION
**Pull request by jace**

### CODE REVIEW

**Comment by jace** [comment]
The terms "state" and "status" appear interchangeable, but common usage shows a slight bias in meanings, which I used when deciding to call this `StateProperty` instead of `StatusProperty`:

1. "State" indicates the state of an item in a process or workflow, without value judgement on the state. It typically indicates where the object is.
2. "Status" indicates _how_ the object is doing. Good, bad, ahead, lagging, etc.

As "state" is a fixed value while "status" is calculated, I've called this a "state property".

**Comment by jace** [comment]
`StateProperty` must support retrieval of the state label, perhaps with a `.label` attribute.

File: coaster/sqlalchemy/properties.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+        :param str name: Name of the new state
+        :param value: Value or group of values of an existing state
+        :param validator: Function that will be called with the host object as a parameter
+        :param class_validator: Function that will be called when the state is queried
+            on the class instead of the instance. Falls back to ``validator`` if not specified
+        """
+        self.states[name] = (value, validator, class_validator)
+
+    def transition(self, from_, to):
```
Comment (shreyas-satish) [inline coaster/sqlalchemy/properties.py]: Is it necessary to mention `from_`? Would there be cases where `from_` may not matter?

Comment (shreyas-satish) [reply coaster/sqlalchemy/properties.py]: I see it allows a list, so I suppose that helps.

**Pull request merged**